compiler/rustc_session/src/errors.rs

```diff
@@ -32,6 +32,22 @@ pub struct FeatureDiagnosticForIssue {
     pub n: NonZeroU32,
 }
 
+#[derive(Subdiagnostic)]
+#[note(session_feature_suggest_upgrade_compiler)]
+pub struct SuggestUpgradeCompiler {
+    version: &'static str,
+    date: &'static str,
+}
+
+impl SuggestUpgradeCompiler {
+    pub fn new() -> Self {
+        let version = option_env!("CFG_VERSION").unwrap_or("unknown");
+        let date = option_env!("CFG_VER_DATE").unwrap_or("unknown");
```

Comment:
Can this just return `-> Option<Self>` and give up if we can't parse the date?

Reply:
In this case should the message just be omitted?

Reply:
Yes

Reply:
I have made this optional depending on the presence of the date environment variable